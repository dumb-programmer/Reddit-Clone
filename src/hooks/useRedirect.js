import { useNavigate } from "react-router-dom"

const useRedirect = (to, message) => {
    const navigate = useNavigate();

    const redirect = () => {
        navigate(to, { state: { message } });
    }

    return redirect;
}

export default useRedirect;