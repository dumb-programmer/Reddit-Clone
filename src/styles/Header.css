.nav-bar {
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 8px;
    border-bottom: 1px solid #edeff1;
    position: fixed;
    top: 0;
    width: 100%;
    background-color: #fff;
    z-index: 5;
}

#search-bar {
    flex: 2;
    border-radius: 19px;
    border: none;
    background-color: #f6f7f8;
    padding: 10px;
    padding-left: 40px;
}

#search-bar:focus, #search-bar:hover {
    outline: 1px solid #0079d3;
}

.user-controls {
    display: flex;
    flex-direction: column;
    justify-content: center;
    position: relative;
    margin-right: 1rem;
}

#login-btn, #sign-up-btn {
    border-radius: 18px;
    width: 100px;
    border: none;
    padding: 5px;
    font-weight: bolder;
    margin-right: 15px;
}

#sign-up-btn {
    color: #0079d3;
    background-color: white;
    border: 0.5px solid;
    border-color: #0079d3;
}

#sign-up-btn:hover {
    background-color: #f4f9fd;
}

#login-btn {
    background-color: #0079d3;
    color: #eee;
    padding: 7px;
}

.user-info {
    cursor: pointer;
    display: flex;
    align-items: center;
    gap: 0.5em;
}

.user-info:hover {
    outline: 1px solid #edeff1;
}

.dropdown {
    position: absolute;
    border: 1px solid #edeff1;
    border-radius: 5px;
    top: 35px;
    width: 100%;
    background-color: #fff;
}

.dropdown>ul {
    padding: 0;
}

.dropdown-link {
    display: flex;
    align-items: center;
    gap: 0.5rem;
    list-style: none;
    padding: 8px;
    padding-left: 20px;
    cursor: pointer;
    text-decoration: none;
    color: inherit;
}

.dropdown-link:hover {
    background-color: #f5f5f5;
}

.dropdown-link:active {
    background-color: #ebebeb;
}