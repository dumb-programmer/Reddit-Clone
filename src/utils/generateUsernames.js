import randomWords from "random-words";

const generateUsernames = () => {
    return Array.from({ length: 5 }).map(
        () => randomWords() + "_" + randomWords() + Math.ceil(Math.random() * 1000)
    );
};

export default generateUsernames;