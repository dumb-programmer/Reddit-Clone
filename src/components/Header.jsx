import { useEffect, useState } from "react";
import { Link, useNavigate } from "react-router-dom";
import { logout, subscribeToUserDoc } from "../firebase";
import useRedirect from "../hooks/useRedirect";
import LogoutIcon from "./icons/LogoutIcon";
import RedditIcon from "./icons/RedditIcon";
import SettingsIcon from "./icons/SettingsIcon";
import UserIcon from "./icons/UserIcon";
import SearchBar from "./SearchBar";
import useAuthContext from "../hooks/useAuthContext";
import "../styles/Header.css";

const Header = () => {
  const [logoutDisabled, setLogoutDisabled] = useState(false);
  const [showDropdown, setShowDropdown] = useState(false);
  const [profilePicture, setProfilePicture] = useState(null);
  const redirectToHome = useRedirect("/");
  const auth = useAuthContext();

  const navigate = useNavigate();

  const handleLogout = async (e) => {
    e.stopPropagation();
    setLogoutDisabled(true);
    await logout();
    localStorage.removeItem("username");
    localStorage.removeItem("user");
    navigate("/");
  };

  useEffect(() => {
    let unsubUser = null;
    if (auth) {
      unsubUser = subscribeToUserDoc(auth.uid, (snap) => {
        setProfilePicture(snap.data().profilePicture);
      });
    }
    return () => {
      unsubUser && unsubUser();
    };
  }, [auth]);

  useEffect(() => {
    window.addEventListener("click", () => {
      if (showDropdown) {
        setShowDropdown(false);
      }
    });
  }, [showDropdown]);

  return (
    <header className="nav-bar">
      <div
        className="logo"
        onClick={redirectToHome}
        style={{ cursor: "pointer" }}
      >
        <RedditIcon height={35} width={35} />
      </div>
      <SearchBar />
      <div className="user-controls">
        {localStorage.getItem("username") ? (
          <>
            <div
              className="user-info"
              styles={{ marginRight: "5px" }}
              onClick={(e) => {
                e.stopPropagation();
                setShowDropdown((prev) => !prev);
              }}
            >
              <img
                src={profilePicture}
                height="30px"
                width="30px"
                alt="user-avatar"
              />
              <span>{localStorage.getItem("username")}</span>
              <svg
                xmlns="http://www.w3.org/2000/svg"
                width="24"
                height="24"
                viewBox="0 0 24 24"
                fill="none"
                stroke="#888b8d"
                strokeWidth="1.25"
                strokeLinecap="round"
                strokeLinejoin="round"
                className="feather feather-chevron-down"
              >
                <polyline points="6 9 12 15 18 9"></polyline>
              </svg>
            </div>
            {showDropdown && (
              <div className="dropdown">
                <ul>
                  <Link
                    to={`/user/${localStorage.getItem("username")}`}
                    className="dropdown-link"
                  >
                    <UserIcon height={20} width={20} />
                    Profile
                  </Link>
                  <Link to={`/settings`} className="dropdown-link">
                    <SettingsIcon height={20} width={20} />
                    Settings
                  </Link>
                  <li
                    className="dropdown-link"
                    onClick={!logoutDisabled ? handleLogout : null}
                  >
                    <LogoutIcon height={20} width={20} />
                    Logout
                  </li>
                </ul>
              </div>
            )}
          </>
        ) : (
          <>
            <button
              className="btn"
              id="sign-up-btn"
              onClick={() => navigate("/signup")}
            >
              Sign Up
            </button>
            <button
              className="btn"
              id="login-btn"
              onClick={() => navigate("/login")}
            >
              Log In
            </button>
          </>
        )}
      </div>
    </header>
  );
};

export default Header;
