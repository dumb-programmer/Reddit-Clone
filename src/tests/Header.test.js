import { screen, render } from "@testing-library/react";
import Header from "../components/Header";
import '@testing-library/jest-dom';
import AuthContext from "../context/AuthContext";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router-dom";
import * as Firebase from "../firebase";
import useRedirect from "../hooks/useRedirect";
import generateUsernames from "../utils/generateUsernames";

jest.mock("../firebase.js");
jest.mock("../hooks/useRedirect", () => {
    const redirect = jest.fn();
    const useRedirect = jest.fn(() => redirect);
    return useRedirect;
});

jest.mock("../utils/generateUsernames", () => {
    return jest.fn(() => ["average_up902", "board_himself128", "contrast_exact736"]);
})

const userDoc = { username: "test", usernameConfirmed: false };
userDoc.data = () => userDoc;

const unsub = jest.fn();
Firebase.subscribeToUserDoc = jest.fn((id, cb) => {
    cb(userDoc);
    return unsub;
})

describe("Header", () => {
    const auth = { uid: "fjadkslfd23423kjsd" };
    const user = userEvent.setup();
    test("Unauthenticated users", () => {
        const { container } = render(<MemoryRouter>
            <Header />
        </MemoryRouter>);
        expect(container).toMatchSnapshot();
    });
    describe("Authenticated users", () => {
        test("Renders", () => {
            const { container, unmount } = render(<MemoryRouter>
                <AuthContext.Provider value={auth}>
                    <Header />
                </AuthContext.Provider>
            </MemoryRouter>);
            expect(container).toMatchSnapshot();
            expect(Firebase.subscribeToUserDoc).toBeCalled();
            unmount();
            expect(unsub).toBeCalled();
        });
        describe("Dropdown works", () => {
            beforeEach(async () => {
                render(<MemoryRouter>
                    <AuthContext.Provider value={auth}>
                        <Header />
                    </AuthContext.Provider>
                </MemoryRouter>);
                await user.click(screen.getByTestId("user-controls"));
            });
            test("Renders", () => {
                expect(screen.getByText(/profile/i)).toBeInTheDocument();
                expect(screen.getByText(/profile/i)).toHaveAttribute("href", `/user/${userDoc.username}`)
                expect(screen.getByText(/settings/i)).toBeInTheDocument();
                expect(screen.getByText(/settings/i)).toHaveAttribute("href", "/settings");
                expect(screen.getByText(/logout/i)).toBeInTheDocument();
            });
            test("Logout button works", async () => {
                await user.click(screen.getByText(/logout/i));
                expect(Firebase.logout).toBeCalled();
            });
        });
        describe("User's that've sign in using auth providers can change their username", () => {

            describe("UsernameConfirmation", () => {
                let keepUsernameBtn;
                let changeUsernameBtn;
                beforeEach(() => {
                    render(<MemoryRouter>
                        <AuthContext.Provider value={auth}>
                            <Header />
                        </AuthContext.Provider>
                    </MemoryRouter>);
                    keepUsernameBtn = screen.getByText(/keep username/i);
                    changeUsernameBtn = screen.getByText(/change username/i);
                });
                test("Renders", () => {
                    expect(screen.getByText(/Do you want to change or keep this user name?/i)).toBeInTheDocument();
                    expect(screen.getByText(`u/${userDoc.username}`)).toBeInTheDocument();
                    expect(keepUsernameBtn).toBeInTheDocument();
                    expect(keepUsernameBtn).toHaveClass("tertiary-btn");
                    expect(changeUsernameBtn).toBeInTheDocument();
                    expect(changeUsernameBtn).toHaveClass("primary-btn");
                });
                describe("Keep username button works", () => {
                    beforeEach(async () => {
                        await user.click(keepUsernameBtn);
                    });
                    test("KeepUsernameModal is shown", async () => {
                        expect(screen.getByText("Are you sure this will be your username forever")).toBeInTheDocument();
                        expect(screen.getByTestId("keep-username-btn-final")).toBeInTheDocument();
                    });
                    test("Modal's keep username button works", async () => {
                        await user.click(screen.getByTestId("keep-username-btn-final"));
                        expect(Firebase.keepUsername).toBeCalled();
                        expect(Firebase.keepUsername).toBeCalledWith(auth.uid);
                    });
                    test("Change username buttons shows the ChangeUsernameModal", async () => {
                        await user.click(screen.getByTestId("change-username-btn-final"));
                        expect(screen.getByText("Can't think of one use one of these")).toBeInTheDocument();
                        expect(screen.getByText(/continue/i)).toBeInTheDocument();
                    });
                });
                describe("Change username button works", () => {
                    let usernameInput;
                    let continueBtn;
                    beforeEach(async () => {
                        await user.click(changeUsernameBtn);
                        usernameInput = screen.getByTestId("username-input");
                        continueBtn = screen.getByText(/continue/i);
                    });
                    test("ChangeUsernameModal is shown", () => {
                        expect(screen.getByText("Can't think of one use one of these")).toBeInTheDocument();
                        expect(usernameInput).toBeInTheDocument();
                        expect(continueBtn).toBeInTheDocument();
                        expect(continueBtn).toBeDisabled();
                    });
                    test("Username input works", async () => {
                        await user.type(usernameInput, "test");
                        expect(usernameInput).toHaveValue("test");
                        expect(continueBtn).not.toBeDisabled();
                    });
                    test("Username suggestions work", async () => {
                        expect(generateUsernames).toBeCalled();
                        const usernames = generateUsernames();
                        usernames.forEach(async (username) => {
                            expect(screen.getByText(username)).toBeInTheDocument();
                        });
                        await user.click(screen.getByText(usernames[0]));
                        expect(usernameInput).toHaveValue(usernames[0]);
                    })
                    test("Continue buttons works", async () => {
                        await user.type(usernameInput, "test");
                        await user.click(continueBtn);
                        expect(Firebase.usernameAvailable).toBeCalled();
                        expect(Firebase.usernameAvailable).toBeCalledWith("test");
                    });
                    test("Username input is validated", async () => {
                        Firebase.usernameAvailable = jest.fn((username) => {
                            if (username === "test") return false;
                            return true;
                        })
                        await user.type(usernameInput, "test");
                        await user.click(continueBtn);
                        expect(Firebase.usernameAvailable).toHaveReturnedWith(false);
                        expect(usernameInput).toHaveClass("form-input__error");
                        expect(screen.getByText(/that username is already taken/i)).toBeInTheDocument();
                        expect(screen.getByText(/that username is already taken/i)).toHaveClass("error-message");
                        expect(continueBtn).toBeDisabled();
                    });
                    test("Form progresses on continue if the username is available", async () => {
                        await user.type(usernameInput, "jester");
                        await user.click(continueBtn);
                        expect(screen.getByText(/are you sure\? this will be your username forever/i)).toBeInTheDocument();
                        expect(screen.getByText("u/jester")).toBeInTheDocument();
                        expect(screen.getByText(/save username/i)).toBeInTheDocument();
                        expect(screen.getByText(/save username/i)).toHaveClass("primary-btn");
                        expect(screen.getByText(/go back/i)).toHaveClass("tertiary-btn");
                    });
                    test("Save username button works", async () => {
                        await user.type(usernameInput, "jester");
                        await user.click(continueBtn);
                        await user.click(screen.getByText(/save username/i));
                        expect(Firebase.changeUsername).toBeCalled();
                        expect(Firebase.changeUsername).toBeCalledWith(auth.uid, "jester");
                    });
                    test("Go back button works", async () => {
                        await user.type(usernameInput, "jester");
                        await user.click(continueBtn);
                        await user.click(screen.getByText(/go back/i));
                        expect(screen.getByTestId("username-input")).toBeInTheDocument();
                        expect(screen.getByText(/continue/i)).toBeInTheDocument();
                    });
                });
            });
        })
    });
    test("Search works", async () => {
        render(<MemoryRouter>
            <AuthContext.Provider value={auth}>
                <Header />
            </AuthContext.Provider>
        </MemoryRouter>);
        const searchBox = screen.getByRole("textbox");
        await user.type(searchBox, "Hello");
        expect(searchBox).toHaveValue("Hello");
        await user.click(searchBox);
        await user.keyboard("[Enter]");
        expect(useRedirect()).toBeCalled();
    });
});