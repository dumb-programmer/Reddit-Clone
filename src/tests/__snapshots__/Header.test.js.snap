// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Authenticated users Renders 1`] = `
<div>
  <header
    class="nav-bar"
  >
    <div
      class="logo"
      style="cursor: pointer;"
    >
      <svg
        class="SVG-icon SnooIcon"
        style="height: 35px; width: 35px;"
        version="1.1"
        viewBox="0 0 20 20"
        x="0px"
        xmlns="http://www.w3.org/2000/svg"
        y="0px"
      >
        <g>
          <circle
            cx="10"
            cy="10"
            fill="#ff4500"
            r="10"
          />
          <path
            d="M16.67,10A1.46,1.46,0,0,0,14.2,9a7.12,7.12,0,0,0-3.85-1.23L11,4.65,13.14,5.1a1,1,0,1,0,.13-0.61L10.82,4a0.31,0.31,0,0,0-.37.24L9.71,7.71a7.14,7.14,0,0,0-3.9,1.23A1.46,1.46,0,1,0,4.2,11.33a2.87,2.87,0,0,0,0,.44c0,2.24,2.61,4.06,5.83,4.06s5.83-1.82,5.83-4.06a2.87,2.87,0,0,0,0-.44A1.46,1.46,0,0,0,16.67,10Zm-10,1a1,1,0,1,1,1,1A1,1,0,0,1,6.67,11Zm5.81,2.75a3.84,3.84,0,0,1-2.47.77,3.84,3.84,0,0,1-2.47-.77,0.27,0.27,0,0,1,.38-0.38A3.27,3.27,0,0,0,10,14a3.28,3.28,0,0,0,2.09-.61A0.27,0.27,0,1,1,12.48,13.79Zm-0.18-1.71a1,1,0,1,1,1-1A1,1,0,0,1,12.29,12.08Z"
            fill="#fff"
          />
        </g>
      </svg>
    </div>
    <div
      style="min-width: 50%; position: relative; display: flex;"
    >
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="height: 20px; width: 20px; stroke: #898c8e; position: absolute; top: 8px; left: 10px;"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <line
          x1="21"
          x2="16.65"
          y1="21"
          y2="16.65"
        />
      </svg>
      <input
        id="search-bar"
        placeholder="Search Reddit"
        type="text"
        value=""
      />
    </div>
    <div
      class="change-username-dropdown"
      style="position: absolute; top: 60px; right: 5px; background-color: rgb(255, 255, 255); z-index: 2; padding: 20px; text-align: center; max-width: 350px; border-radius: 5px;"
    >
      <p>
        Do you want to change or keep this user name?
      </p>
      <h2>
        u/
        test
      </h2>
      <div
        style="display: flex; flex-direction: column; gap: 1rem; padding: 0px 20px;"
      >
        <button
          class="primary-btn"
        >
          Change Username
        </button>
        <button
          class="tertiary-btn"
        >
          Keep Username
        </button>
      </div>
    </div>
    <div
      class="user-controls"
      data-testid="user-controls"
    >
      <div
        class="user-info"
        styles="[object Object]"
      >
        <img
          alt="user-avatar"
          height="30px"
          width="30px"
        />
        <span
          class="username"
        >
          test
        </span>
        <svg
          class="feather feather-chevron-down"
          fill="none"
          height="24"
          stroke="#888b8d"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="1.25"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header Unauthenticated users 1`] = `
<div>
  <header
    class="nav-bar"
  >
    <div
      class="logo"
      style="cursor: pointer;"
    >
      <svg
        class="SVG-icon SnooIcon"
        style="height: 35px; width: 35px;"
        version="1.1"
        viewBox="0 0 20 20"
        x="0px"
        xmlns="http://www.w3.org/2000/svg"
        y="0px"
      >
        <g>
          <circle
            cx="10"
            cy="10"
            fill="#ff4500"
            r="10"
          />
          <path
            d="M16.67,10A1.46,1.46,0,0,0,14.2,9a7.12,7.12,0,0,0-3.85-1.23L11,4.65,13.14,5.1a1,1,0,1,0,.13-0.61L10.82,4a0.31,0.31,0,0,0-.37.24L9.71,7.71a7.14,7.14,0,0,0-3.9,1.23A1.46,1.46,0,1,0,4.2,11.33a2.87,2.87,0,0,0,0,.44c0,2.24,2.61,4.06,5.83,4.06s5.83-1.82,5.83-4.06a2.87,2.87,0,0,0,0-.44A1.46,1.46,0,0,0,16.67,10Zm-10,1a1,1,0,1,1,1,1A1,1,0,0,1,6.67,11Zm5.81,2.75a3.84,3.84,0,0,1-2.47.77,3.84,3.84,0,0,1-2.47-.77,0.27,0.27,0,0,1,.38-0.38A3.27,3.27,0,0,0,10,14a3.28,3.28,0,0,0,2.09-.61A0.27,0.27,0,1,1,12.48,13.79Zm-0.18-1.71a1,1,0,1,1,1-1A1,1,0,0,1,12.29,12.08Z"
            fill="#fff"
          />
        </g>
      </svg>
    </div>
    <div
      style="min-width: 50%; position: relative; display: flex;"
    >
      <svg
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        style="height: 20px; width: 20px; stroke: #898c8e; position: absolute; top: 8px; left: 10px;"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <line
          x1="21"
          x2="16.65"
          y1="21"
          y2="16.65"
        />
      </svg>
      <input
        id="search-bar"
        placeholder="Search Reddit"
        type="text"
        value=""
      />
    </div>
    <div>
       
      <button
        class="btn"
        id="sign-up-btn"
      >
        Sign Up
      </button>
      <button
        class="btn"
        id="login-btn"
      >
        Log In
      </button>
    </div>
  </header>
</div>
`;
